README.md
# dojo-cli

A command-line interface library for black box domain model execution. This library enables uses to execute domain models locally.

The library has 6 key methods:

- List the latest versions of models.
- Get parameter metadata for a selected model.
- Run a model via the Docker Python SDK.
- Check model container status.
- Get model (docker container) logs.
- Get results path.


## Setup

The CLI requires a configuration file with Docker Hub and DOJO credentials. This filename can be passed with each CLI command, or the default file `.config` will be used.

See `example.config` for guidance:
```
{
    "DOCKERHUB_USER": "",
    "DOCKERHUB_PWD": "",
    "DOJO_URL": "https://dojo-test.com",
    "DOJO_USER": "",
    "DOJO_PWD": ""
}
```

## Running the CLI from source


### CLI help:
```
python dojocli/cli.py --help
python dojocli/cli.py get-model-params --help
python dojocli/cli.py list-models --help
python dojocli/cli.py run-model --help
```

### List models

$ `python dojocli/cli.py list-models` lists the available models e.g.: 
```
(1) APSIM
(2) APSIM-Cropping
(3) APSIM-Rangelands
(4) Accessibility Model
(5) AgMIP Seasonal Crop Emulator
(6) CHIRPS - Climate Hazards Center Infrared Precipitation with Stations
(7) CHIRPS-GEFS
(8) CHIRPS-GEFS Monthly
(9) CHIRPS-Monthly
...
```

### Get the run parameters of a model

$ `python dojocli/cli.py get-model-params --model_name=CHIRPS-Monthly`

This is required before running a model. It will write the parameters for the model CHIRPS-Monthly to `CHIRPS-Monthly_params.txt`:

```
# Lines starting with # are comments.
# Model run parameters for CHIRPS-Monthly
# Example parameters:
# month: 01
# year: 2021
# bbox: [[33.512234, 2.719907], [49.98171,16.501768]]
#
# Set run parameter values here:
name: 
month: 
year: 
bbox: 
```

To run a model, the parameter values above need to be set, and the parameters filename passed to `run_model`.

### Run a model

$ `python dojocli/cli.py run-model --model_name=CHIRPS-Monthly --params=CHIRPS-Monthly_params.txt`

This command will download and run the model docker image.


HISTORY.md
# History

## 0.1.0 (2021-12-25)

-   First release on PyPI.



dojocli/dojo_client.py
"""
  Dojo-cli client for dojo api.
"""

from docker_client import DockerClient
import json
import requests

class DojoClient(object):

    def __init__(self, config_filename):
        self.set_config(config_filename)
        self.image_name = None
        self.model_id = None

    def generic_dojo_get_request(self, url):
        try:
            response = requests.get(url, auth=(self.dojo_auth))
            return response
        except Exception as e:
            if hasattr(e, 'message'):
                print(e.message)
            else:
                print(e)
            exit

    def get_available_models(self):
        """
        Description
        -----------

            Get latest models with images from dojo api.

            Use /models and not models/latest so that images can be restricted to 
            jataware/dojo-publish.

        """

        url = f'{self.dojo_url}/models?query=image:"jataware/dojo" AND NOT _exists_:"next_version"&size=1000'
        response = self.generic_dojo_get_request(url)
        try:
            resp = response.json()
            models = set()
            for r in resp["results"]:
                if r['image'] != '' and r['next_version'] == None:
                    models.add(r['name'])

            models = list(models)
            models.sort()
            return models
        except Exception as e:
            if hasattr(e, 'message'):
                print(e.message + '\n' + response.text)
            else:
                print(e + '\n' + response.text)
            exit

    def get_dojo_stuff(self, stuff, model_id):
        """
            Get stuff(config, directive, accessories, outputfiles) based on model_id.
        """

        url = f'{self.dojo_url}/dojo/{stuff}/{model_id}'
        response = self.generic_dojo_get_request(url)
        try:
            return response.json()
        except Exception as e:
            if hasattr(e, 'message'):
                print(e.message)
            else:
                print(e)
            exit

    def get_metadata(self, model_name: str):
        """
        Description
        -----------
        Calls /dojo endpoints to download model metadata e.g. directive, config, 
        outputfiles, and accessory files.

        Process
        -------
        (1) take the model name and calls GET /models/latest to 
        find the model_id.

        (2) use the model_id to call GET /dojo/ endpoints to get the directive,
        outputfiles, accessories and config.

        (3) return these as json.

        Parameters
        ----------
        model_name: str
            The name of the model of interest.

        Returns
        -------
            dict: metadata as JSON

        """

        # Get the model_id from the model_name
        self.model_id, self.image_name = self.get_model_info(model_name)
        print(f"Model id for {model_name} is {self.model_id}")

        # Get "dojo_stuff", i.e. the four types of information available by GET
        # /dojo/, and add to the metadata.
        metadata = {}
        dojo_stuff = ['accessories', 'config', 'directive', 'outputfile']
        for stuff in dojo_stuff:
            metadata[stuff] = self.get_dojo_stuff(stuff, self.model_id)
        
        return metadata

    def get_model_info(self, model_name):
        """
            Get the model_id and image_name based on the model_name. 
            Restrict query to models with a jataware image and "next_version":null.

        Returns
        -------
            tuple: (model_id, image_name)
        """

        url = f'{self.dojo_url}/models?query=name:"{model_name}" AND image:"jataware/dojo" AND NOT _exists_:"next_version"'
        response = self.generic_dojo_get_request(url)
        try:
            resp = response.json()

            # TODO: multiple hits for the latest model_name: the image appears to be same, though.
            print(resp["hits"], ' get_model_id() hits')
            for r in resp["results"]:
                print(r["image"])

            for r in resp["results"]:
                if r["name"] == model_name:
                    return (r["id"], r["image"])

        except Exception as e:
            if hasattr(e, 'message'):
                print(e.message)
            else:
                print(e)
            exit

    def run_model(self, model_name, params_filename):

        # Get the metadata for this model. This sets self.image_name.
        metadata = self.get_metadata(model_name)
        params = self.set_run_params(params_filename)

        # set run command from metadata["directive"]["command"] and substitute params.

        dc = DockerClient(self.dockerhub_user, self.dockerhub_pwd)
        
        # pull the image
        dc.pull_image(self.image_name)

        # create the container
        container_name = dc.create_container(self.image_name)
        print(f"Created container {container_name}")

        


        

    def set_config(self, config_filename):
        with open(config_filename) as f:
            config = json.load(f)

            # Set dojo url and authentication credentials.
            self.dojo_auth = (config["DOJO_USER"], config["DOJO_PWD"]) 
            self.dojo_url = config["DOJO_URL"]
            
            # Set DockerHub url and credentials.
            self.dockerhub_pwd = config["DOCKERHUB_PWD"]
            self.dockerhub_user = config["DOCKERHUB_USER"]

    def set_run_params(self, params_filename):
        params = {}
        with open(params_filename, 'r') as f:
            for l in f.readlines():
                if not l.startswith("#"):
                    sa = l.split(":")
                    params[sa[0]] = str(sa[1]).strip()
        return params

    def test_docker(self):
        dc = DockerClient(self.dockerhub_user, self.dockerhub_pwd)
        #dc.pull_image('jataware/dojo-publish', 'CHIRPS-Monthly-latest')
        dc.list_system_images()

dojocli/docker_client.py


import docker
import json

from docker import client

class DockerClient(object):

    def __init__(self, user, pwd):
        self.client = docker.from_env()
        self.auth_config = { "username": user, "password": pwd}
        self.container = None

    def list_system_images(self):
        for i in self.client.images.list(name='jataware/dojo-publish'):
            print(i)

    def pull_image(self, image_name):
        """
        Description
        -----------
        Pull the image from hub.docker.com to the local docker system.
        
        Parameters
        ----------
            imagename: str
                format repo:tag
                e.g. jataware/dojo-publish:CHIRPS-Monthly-latest
        """

        sa = image_name.split(":")
        repo = sa[0]
        tag = sa[1]
        for line in self.client.api.pull(repo, tag, True, self.auth_config, decode=True):
            print(json.dumps(line, indent=4))
            # TODO: Progress Bar
               

    def create_container(self, image_name):
        self.container = self.client.containers.run(image_name, command='bash', stdin_open=True, detach=True)
        return self.container.name










dojocli/cli.py
"""
"""

import click
from dojo_client import DojoClient




def generate_model_params_file(metadata, model_name):
    """
    Generate and write to file a list of model parameters and instructions 
    derived from the directive portion of the metadata.
    """

    params_output = []
    param_list = []
    directive = metadata['directive']

    params_output.append("# Lines starting with # are comments.")

    # Parse and add the command parameters.
    # "command_raw": "python3 run_chirps_tiff.py --name=CHIRPS --month=01 --year=2021 --bbox='[[33.512234, 2.719907], [49.98171,16.501768]]'",
    params_output.append(f'# Model run parameters for {model_name}')
    params_output.append('# Example parameters:')
    command_raw = directive["command_raw"]
    param_array = command_raw.split("--")
    for param_pair in param_array:
        if str(param_pair).lower().startswith('python'):
            continue

        # split the param_pair
        sa = param_pair.split("=")
        
        # keep a list of the parameter names
        param_name = sa[0]
        param_list.append(param_name)

        param_example = str(sa[1]).strip()

        if param_name == 'name':
            continue

        # strip the quotes from string parameters
        if param_example[0] == "'":
            param_example = param_example[1:-1]

        # add the example param:value pair.
        params_output.append(f'# {param_name}: {param_example}')

    params_output.append('#\n# Set run parameter values here:')
    for p in param_list:
        params_output.append(f'{p}: ')

    with open(f'{model_name}_params.txt', 'w') as f:
        f.writelines("%s\n" % line for line in params_output)

@click.group()
def cli():
    pass

@cli.command()
@click.option("--model_name", type=str, default=None, help="the model name e.g. CHIRPS-Monthly")
@click.option("--config", type=str, default=".config", help="configuration json filename e.g. sample.config")
def get_model_params(model_name, config):
    """Get the parameters required to run a model."""
    click.echo(f"Getting parameters for {model_name} ...")
    
    dc = DojoClient(config)
    metadata = dc.get_metadata(model_name)
    generate_model_params_file(metadata, model_name)

    click.echo(f"{model_name} parameters file written to {model_name}_params.txt.")

@cli.command()
@click.option("--config", type=str, default=".config", help="configuration json filename e.g. sample.config")
def list_models(config):
    """List available models."""
    click.echo("Listing available models...\n")

    dc = DojoClient(config)

    models = dc.get_available_models() 
    for idx, m in enumerate(models):
        click.echo(f'({idx+1}) {m}')


@cli.command()
@click.option("--model_name", type=str, default=None, help="the model name e.g. CHIRPS-Monthly")
@click.option("--config", type=str, default=".config", help="configuration json filename e.g. sample.config")
@click.option("--params", type=str, default="params.txt", help="model run parameters filename e.g. params.txt")
def run_model(model_name, config, params):
    """Run a model."""
    click.echo(f"Running {model_name}")

    dc = DojoClient(config)
    dc.run_model(model_name, params)

if __name__ == "__main__":     
    cli()
    #get_parameters(model_name="CHIRPS-Monthly", config=".config")
    #run_model(model_name="CHIRPS-Monthly", config=".config", params="CHIRPS-Monthly_params.txt")